Add description for enabling experimental features (#2546)
nginxinc/nginx-gateway-fabric: site/content/includes/installation/install-gateway-api-experimental-features.md

---
docs: "DOCS-000"
---

To use Gateway API experimental resources, the Gateway API resources from the experimental channel must be installed before deploying NGINX Gateway Fabric. Additionally, NGINX Gateway Fabric must have experimental features enabled.

> {{<caution>}}As noted in the [Gateway API documentation](https://gateway-api.sigs.k8s.io/guides/#install-experimental-channel), future releases of the Gateway API can include breaking changes to experimental resources and fields.{{</caution>}}

Author: @ADubhlaoich I see that its a bit inconsistent throughout our current documentation https://github.com/nginxinc/nginx-gateway-fabric/blob/b8c64df41a709ea84e4ecce46ffb9b91ee0a007a/site/content/includes/installation/install-gateway-api-resources.md?plain=1#L5 as an example, do you think this would be something worth standardizing and fixing in our documentation?

> {{<caution>}}As noted in the [Gateway API documentation](https://gateway-api.sigs.k8s.io/guides/#install-experimental-channel), future releases of the Gateway API can include breaking changes to experimental resources and fields.{{</caution>}}

Reviewer: It's not critical so I wouldn't dedicate time to it alone: for those kinds of small changes, I favour making them if I happen to be updating a given page. It displays the same to the user, so it's purely an internal style concern.

To install the Gateway API resources from the experimental channel, run the following:

```shell
kubectl kustomize "https://github.com/nginxinc/nginx-gateway-fabric/config/crd/gateway-api/experimental?ref=v1.4.0" | kubectl apply -f -
```

> 

Reviewer: Could be nice to show example output as the consequence of the `kubectl` command.

Author: I'm up for it, though do you think it would be beneficial to add it to https://github.com/nginxinc/nginx-gateway-fabric/blob/b8c64df41a709ea84e4ecce46ffb9b91ee0a007a/site/content/includes/installation/install-gateway-api-resources.md?plain=1#L10 also? I don't think it makes sense to do it to one but not the other.

Reviewer: Yeah, sounds good to me. It's a pattern (Showing users expected output) we're encouraging adoption of: it should make successful steps clearer for a reader with less experience.

Reviewer 2: Generally, I'm a fan of including output for kubectl commands, but in this case, I don't think we want to. The output is a list of CRDs that are installed for the Gateway API version and channel in the command:

```
customresourcedefinition.apiextensions.k8s.io/gatewayclasses.gateway.networking.k8s.io configured
customresourcedefinition.apiextensions.k8s.io/gateways.gateway.networking.k8s.io configured
customresourcedefinition.apiextensions.k8s.io/grpcroutes.gateway.networking.k8s.io configured
customresourcedefinition.apiextensions.k8s.io/httproutes.gateway.networking.k8s.io configured
customresourcedefinition.apiextensions.k8s.io/referencegrants.gateway.networking.k8s.io configured
```

This list is likely to change with every release of the Gateway API. I don't think the benefit to the user is big enough to warrant maintaining this output.

{{<note>}}If you plan to use the `edge` version of NGINX Gateway Fabric, you can replace the version in `ref` with `main`, for example `ref=main`.{{</note>}}

To enable experimental features on NGINX Gateway Fabric:

> If using Helm: The `nginxGateway.gwAPIExperimentalFeatures.enable` option must be set to true. An example can be found
> in the [Installation with Helm]({{< relref "installation/installing-ngf/helm.md#custom-installation-options" >}}) guide.
> 
> If using Kubernetes manifests: The `--gateway-api-experimental-features` command-line flag must be set to true on the deployment manifest.
> An example can be found in the [Installation with Kubernetes manifests]({{< relref "installation/installing-ngf/manifests.md#3-deploy-nginx-gateway-fabric" >}}) guide.

Reviewer: ```suggestion
Using Helm: Set `nginxGateway.gwAPIExperimentalFeatures.enable` to true. An example can be found
in the [Installation with Helm]({{< relref "installation/installing-ngf/helm.md#custom-installation-options" >}}) guide.

Using Manifests: Add the  `--gateway-api-experimental-features` command-line flag to the deployment manifest.
An example can be found in the [Installation with Kubernetes manifests]({{< relref "installation/installing-ngf/manifests.md#3-deploy-nginx-gateway-fabric" >}}) guide.
```

Changes for conciseness. The Kubernetes instruction might be wrong: I assume that the argument is implicitly true, and doesn't have additional parameters.

Author: Made the change though its slightly adjusted, your assumption is correct.

Reviewer: Fair enough! In context there's no other kind of Manifest it could be mistaken for, so I favour the shorter noun.

The specificity is more important for components with more generic or shared names.

---
docs: "DOCS-000"
---

> Installing Gateway API resources from the experimental channel includes everything in the standard release channel plus additional experimental resources and fields.

Author: So would you be fine changing the description to just be 

> To use Gateway API experimental resources, the Gateway API resources from the experimental channel must be installed before deploying NGINX Gateway Fabric. Additionally, NGINX Gateway Fabric must have experimental features enabled.

I included the first couple of lines to add a little background to experimental features as I feel like without it, it feels lacking information.

> Installing Gateway API resources from the experimental channel includes everything in the standard release channel plus additional experimental resources and fields.

Reviewer 2: Yes, we don't need a lot of information here. It's not a standalone doc, just a doc that explains how to use the experimental resources with NGF. 

Lines 5-7 would be a good addition to the installation docs, especially where we link to the gateway compat doc. That way, when the user installs NGF, they can click on the gateway compat doc and see if they need to install the experimental CRDs based on our chart.

> Installing Gateway API resources from the experimental channel includes everything in the standard release channel plus additional experimental resources and fields.

Author: I moved it to the includes part of the install gateway api resources document. let me know of your thoughts!

> NGINX Gateway Fabric currently supports a subset of the additional features provided by the experimental channel. In order

Author: Thoughts on linking to https://docs.nginx.com/nginx-gateway-fabric/overview/gateway-api-compatibility/ to show which additional features NGF supports?

Reviewer 2: I like it!

To learn more about what Gateway API resources NGINX Gateway Fabric currently supports, visit our [Gateway API Compatibility]({{< relref "overview/gateway-api-compatibility.md" >}}) document.

> In order to use these experimental resources, the [Gateway API resources](https://github.com/kubernetes-sigs/gateway-api) from the experimental channel must be installed before deploying NGINX Gateway Fabric.

Reviewer 2: I don't think this link is needed.
